README.md
# Search Bookmarks, History and Browser Tabs

🔎 Browser extension to (fuzzy) search and navigate bookmarks, history and open tabs.

## Installation via Store

- [Chrome Extension](https://chrome.google.com/webstore/detail/tabs-bookmark-and-history/cofpegcepiccpobikjoddpmmocficdjj?hl=en-GB&authuser=0)
- [Firefox Addon](https://addons.mozilla.org/en-US/firefox/addon/search-tabs-bookmarks-history/)
- [Microsoft Edge Addon](https://microsoftedge.microsoft.com/addons/detail/search-tabs-bookmarks-an/ldmbegkendnchhjppahaadhhakgfbfpo)
- [Opera Addon](https://addons.opera.com/en/extensions/details/search-bookmarks-history-and-tabs/)

## Features

**This extension does not collect any data nor does it make any external requests** (see [Privacy](#privacy--data-protection)).

It supports three different search approaches:

- **Exact search** (case-insensitive, but exact matching): Faster, but only exact matching results.
- **Fuzzy search** (approximate matching): Slower, but includes also inexact (fuzzy) matches.

With this extension you can also **tag your bookmarks** including auto completions.
The tags are considered when searching and can be used for navigation.

The extension is very customizable (see [user options](#user-configuration)) and has a dark / light theme that is selected based on your system settings (see [prefers-color-scheme](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme)).

For a list of recent changes, see [CHANGELOG.md](./CHANGELOG.md).

## Screenshots & Demo

![Demo Animation](/images/bookmark-and-history-search.gif 'Demo Animation')

## Privacy / Data Protection

This extension is built to respect your privacy:

- It does not have permissions for outside communication, so none of your data is shared or exposed externally.
- The extension does not even store any information except your user settings.
  Every time the extension popup is closed, it "forgets" everything and starts from a blank slate next time you open it.
- There is no background job / processing. If the popup is not explicitly opened by the user, the extension is not executed.
- The extension only requests the following permissions for the given reasons:
  - **bookmarks**: Necessary to read and edit the bookmarks. Can be disabled via [user configuration](#user-configuration).
  - **history**: Necessary to read the browsing history. Can be disabled or limited via [user configuration](#user-configuration).
  - **tabs**: Necessary to find open tabs and to use tabs for navigation. Can be disabled via [user configuration](#user-configuration).
  - **storage**: Necessary to store and retrieve the [user configuration](#user-configuration).
    If the browser has setting synchronization enabled, the extension settings will be synced (in this case you already trust your browser to sync everything else anyway).
    If browser sync is disabled, the user configuration is only stored locally.
- The extension is open source, so feel free to convince yourself :)

## User Documentation

- **Search Strategies**: Switch between precise and fuzzy approach by clicking on the FUZZY or PRECISE button in the search bar (top right).
- **Keyboard Shortcut**: Trigger the extension via keyboard.
  - The default is `CTRL` + `Shift` + `.`, but you can customize this.
- **Open selected results**: By default, the extension will open the selected result in a new active tab, or switch to an existing tab with the target url.
  - Hold `Shift` or `Alt` to open the result in the current tab
  - Hold `Ctrl` to open the result without closing the popup.
- **Search Modes**: In case you want to be more selective -> use a search mode:
  - Start your query with `t ` (including space): only **tabs** will be searched.
  - Start your query with `b ` (including space): only **bookmarks** will be searched.
  - Start your query with `h ` (including space): only **history** will be searched.
  - Start your query with `s ` (including space): only **search engines** will be proposed.
  - Start your query with `#`: only **bookmarks with the tag** will be returned (exact "starts with" search)
    - Add `  ` (double space) or press TAB to further search the results
    - Supports AND search, e.g. search for `#github #pr` to only get results which have both tags
  - Start your query with `~`: only **bookmarks within the folder** will be returned (exact "starts with" search)
    - Add `  ` (double space) or press TAB to further search the results
    - Supports AND search, e.g. search for `~Sites ~Blogs` to only get results which have both tags
  - Custom Aliases:
    - The option `customSearchEngines` allows you to define your own search mode aliases
    - Default: Start your query with `g ` (including space): Do a Google search.
    - Default: Start your query with `d ` (including space): Do a dict.cc search.
- **Special Browser Pages**: You can add special browser pages to your bookmarks, like `chrome://downloads`.
- **Custom Scores**: Add custom bonus scores by putting ` +<whole number>` to your bookmark title (before tags)
  - Examples: `Bookmark Title +20` or `Another Bookmark +10 #tag1 #tag2`
- This extension works best if you avoid:
  - using `#` in bookmark titles that do not indicate a tag.
  - using `~` in bookmark folder names.
  - using ` ` (spaces) in tag names (ideally also folder names)

## User Configuration

The extension is highly customizable.
Finding and setting options is a bit technical, though.

The user options are written in [YAML](https://en.wikipedia.org/wiki/YAML) or [JSON](https://en.wikipedia.org/wiki/JSON) notation.

For now, there is no nice options overview, so you have to find them in the [popup/js/model/options.js](popup/js/model/options.js) file in the `defaultOptions` object.
From there you can see the available options, their names, default values and descriptions.

When defining your custom config, you only need to define the options that you want to overwrite from the defaults.

> ⚠ The options are not validated properly. Please make sure to use them correctly.<br/>
> If something breaks, consider resetting your options.

An exemplary user config can look like the following example:

```yaml
searchStrategy: fuzzy
displayVisitCounter: true
displayDateAdded: true
```

If you have troubles with performance, here are a few options that might help. Feel free to pick & choose and tune the values to your situation:

```yaml
searchStrategy: precise
searchMinMatchCharLength: 2
displaySearchMatchHighlight: false,
searchMaxResults: 20
historyMaxItems: 200
```

> Precise search is faster than fuzzy search.
> Start searching only when at least 2 characters are entered
> Highlighting search matches takes some effort, this can be disabled.
> Number of search results can be further limited
> Number of browser history items can be further reduced

Or a more advanced example:

```yaml
searchStrategy: precise
historyDaysAgo: 14
historyMaxItems: 1200
historyIgnoreList:
  - http://localhost
  - http://127.0.0.1
colorStripeWidth: 4 # Customize width of search result color stripe
bookmarkColor: '#46e6e6' # customize color for bookmark results
scoreTabBaseScore: 70 # customize base score for open tabs
searchEngineChoices:
  - name: Google
    urlPrefix: https://google.com/search?q=
customSearchEngines:
  - alias: ['g', 'google']
    name: Google
    urlPrefix: https://www.google.com/search?q=$s
    blank: https://www.google.com
  - alias: d
    name: dict.cc
    urlPrefix: https://www.dict.cc/?s=$s
  - alias: [gh, github]
    name: GitHub
    urlPrefix: https://github.com/search?q=$s
    blank: https://github.com
  - alias: npm
    name: NPM
    urlPrefix: https://www.npmjs.com/search?q=$s
    blank: https://www.npmjs.com
```

In case of making multilingual searching (CJK) correctly, you may need to tweak [uFuzzy](https://github.com/leeoniya/uFuzzy) options via option `ufuzzyOptions`, for example:

```yaml
# make CJK chars work for fuzzy search
uFuzzyOptions:
  interSplit: (p{Unified_Ideograph=yes})+
```

## Scoring System

The scoring systems works roughly the following:

- Depending on the type of result (bookmark, tab, history) a different base score is taken (e.g. `scoreBookmarkBaseScore`).
- Depending on in which result field (title, url, tag, folder) the match was found, the search match gets weighted by multiplication. (e.g. `scoreTitleWeight`).
- This base score is now merged / multiplied with the search library score. A less good match will usually reduce the score and a perfect / highest ranked match will keep it at .
- Depending on certain conditions some bonus score points are added on top. For example, `exactStartsWithBonus` will add score if either the title or the url start exactly with the search term, including spaces.

For a description of the scoring options and what they do, please see [popup/js/model/options.js](popup/js/model/options.js).

## Local Development

### Install and Build

Prerequisite: [Node.js](https://nodejs.org/en/)

```sh
# install dependencies
npm install

# build extension
npm run build
```

The source code for the extension can be found in [popup/](popup/) (HTML, JS and libs) and [sass/](sass/) (SCSS/CSS).

The built extensions can be found

- [dist/chrome/](dist/chrome/) for Google Chrome and Microsoft Edge
- [dist/firefox/](dist/firefox/) for Firefox
- [dist/opera/](dist/firefox/) for Opera

### Developer Installation

- Check out this extension via git
- Run `npm install` and `npm run build` (via bash / git bash)
- **For Chrome / Edge**:
  - Navigate to extensions page (`chrome://extensions/` on Chrome and `edge://extensions/` on Edge).
  - Enable "Developer mode"
  - Choose "Load unpacked" and open the root folder of this repository
- **For Firefox**:

  - First [install and build](#install-and-build) this project.
  - Load the built extension in `dist/firefox` as a temporary addon in `about:debugging`.

### Developer Workflow

- Typical developer actions
  - `npm run build` for a complete build
  - `npm run start` to start the extension locally in a browser (with mock data)
  - `npm run test` to run end to end tests
  - for more, see `npm run`

## Credits

This extension makes use of the following helpful open-source projects (thanks!):

- https://github.com/leeoniya/uFuzzy for the fuzzy search algorithm
- https://github.com/yairEO/tagify for the tag autocomplete widget
- https://markjs.io/ for highlighting search matches
- https://www.npmjs.com/package/js-yaml for the user options parsing
- https://bulma.io/ for some minimal CSS base styling
- https://github.com/tabler/tabler-icons for icons

## Feedback and Ideas

> Please create a [GitHub issue](https://github.com/Fannon/search-tabs-bookmarks-and-history/issues) to give your feedback.
> All ideas, suggestions or bug reports are welcome.


CHANGELOG.md
# CHANGELOG

## [unreleased]

## [v1.10.0]

- **NEW**: Search within a bookmark folder or tag:
  - Now it is possible to further search
- **FIXED**: If `debug: true` has been set, the extension crashed on a performance measurement analysis

## [v1.9.1]

- **IMPROVED**: Better handling of long bookmark titles. Some titles are now abbreviated, especially if they are a subpart of the URL anyway.

## [v1.9.0]

- **FIXED**: Fuzzy search now also works with non-ASCII characters like CKJ chars by default
- **FIXED**: Option `bookmarkColor` now also applies to the bookmark folder badge in the search results
- **REMOVED**: Removed hybrid search as the benefits / differences against new fuzzy search are negligible.

## [v1.8.7]

- **NEW**: New option `uFuzzyOptions` that allows to configure the fuzzy search library used by this extension
  - This can be used, e.g. that fuzzy search supports CKJ characters (see README)

## [v1.8.5]

- **CHANGED**: Extension does not log or do performance measurements by default.
  - If you want to enable this, use the option: `debug: true`

## [v1.8.4]

- **NEW**: Added options to configure the color and the width of the color stripe of search results
  - `colorStripeWidth` to set the width
  - `bookmarkColor` and similar to set the color (expressed as CSS color)
- **NEW**: When hovering over an URL, the full URL is displayed as a hover. Requested via [#74](https://github.com/Fannon/search-bookmarks-history-and-tabs/issues/74)
- **IMPROVED**: Updated dependencies

## [v1.8.3]

- **FIXED**: When navigating result items via arrow up, the search text input box curser moved to the beginning of the search string
  - Fix contributed by [@c0001](https://github.com/c0001) in [PR #71](https://github.com/Fannon/search-bookmarks-history-and-tabs/pull/71). Thanks!

## [v1.8.2]

- **FIXED**: Used wrong icon (edit) for closing tabs
- **IMPROVED**: Hover over edit and close icon is now indicated

## [v1.8.1]

- **FIXED**: Missing icon for closing open tabs
  - Tip: This is especially useful if you enter tab search mode via searching `t `
- **FIXED**: Improved very buggy logic to close tabs and update search results

## [v1.8.0]

- **NEW**: Allow definition of custom search engines that are triggered by custom alias
  - New option: `customSearchEngines`, with one default entry: `g ` for executing a google search.
  - Also added `blank` option when no search string is given
  - Allowing for multiple aliases, if defined as an array (`['alias1', 'alias`]`)

## [v1.7.0]

- **CHANGED**: Replaced fuzzy search library fuzzysort with [uFuzzy](https://github.com/leeoniya/uFuzzy)
  - This should give more reliable and relevant search results
  - Performance is a bit better in most cases
  - Related Issue: https://github.com/Fannon/search-bookmarks-history-and-tabs/issues/60
- **IMPROVED**: Performance improvement for both precise and fuzzy search
  - Interim search results are now cached so the search haystack gets smaller when search term is only expanded
- **FIXED**: Custom Bonus Score was sometimes shown in search result title.

## [v1.6.3]

- **CHANGED**: No bundling of external libraries into vendor.min.js as Firefox store objected to this

## [v1.6.1]

- **IMPROVED**: Improved error handling
  - Errors and warnings are now displayed in a dedicated overlay, with more space to show complete context
  - Error handling is more robust, e.g. when failing to load user settings we now display error but fall back to default options

## [v1.6.0]

- **NEW**: Removed indexing phase entirely, which leads to faster load times
- **NEW**: Added close tab functionality
  - See PR [#38](https://github.com/Fannon/search-bookmarks-history-and-tabs/pull/38)
- **CHANGED**: Removed flexsearch dependency and implemented simpler 'precise' search
  - With this change, the new precise search does not index anymore and therefore the extension is quicker to load
  - The search performance got a bit worse (it's a tradeoff), but is still fast enough
- **CHANGED**: Removed fuse.js for fuzzy search and replaced it with https://github.com/farzher/fuzzysort, which requires no indexing
- **IMPROVED**: Code cleanups and better minimized output, which makes the extension smaller.

## [v1.5.0]

- **NEW**: Added hybrid search approach
  - This combines precise and fuzzy search results
  - This also combines the indexing and search time of both approaches, so be aware of the performance impact.
  - Added two new options to weight precise vs. fuzzy search matches
    - `scoreHybridPreciseBonus`
    - `scoreHybridFuzzyBonus`

## [v1.4.0]

- **NEW**: Bookmark edit now supports:
  - Deleting bookmarks through popup
  - Editing the bookmark URL
- **NEW**: Added support for OR tag and folder search
  - Supports queries like `#github #pr` and `~Sites ~Blogs`
- **Improved**: Search behavior around `scoreExactIncludesBonus`
  - Introduced new option `scoreExactIncludesBonusMinChars` which introduces a minimum character match for above option
- **Improved**: behavior of `scoreExactTagMatchBonus` and `scoreExactFolderMatchBonus` to make it work in more situations
- **NEW**: Rudimentary user option validation, to ensure it at least is a proper YAML / JSON object
- **REMOVED**: Legacy option migration

## [v1.3.0]

- Simplified the user options
  - The options can now be written in YAML or JSON (instead of JSON5 / JSON before)
  - The structure is now flat, so there is no need for object nesting except for advanced options
  - Added inline help to option screen

## v1.2.1

- Added inline tips to popup [#27](https://github.com/Fannon/search-bookmarks-history-and-tabs/pull/27).
- New option: `tabs.displayLastVisited` allows to display the last accessed tabs in the default results when popup is opened [#22](https://github.com/Fannon/search-bookmarks-history-and-tabs/pull/22).

## v1.2.0

- NEW: Results can now be opened in current tab or background tab [#18](https://github.com/Fannon/search-bookmarks-history-and-tabs/pull/18).

popup/js/model/options.js
import { printError } from '../helper/utils.js'

/**
 * The default options
 *
 * They can be selectively overwritten and customized via user options
 * @see https://github.com/Fannon/search-tabs-bookmarks-and-history#user-configuration
 */
export const defaultOptions = {
  //////////////////////////////////////////
  // GENERAL OPTIONS                      //
  //////////////////////////////////////////

  /**
   * When enabled, the extension will log some additional information like what it indexed
   * and performance measurements
   */
  debug: false,

  //////////////////////////////////////////
  // SEARCH OPTIONS                       //
  //////////////////////////////////////////

  /**
   * Search approach to use. Choose between:
   *
   * * 'precise': Simple search approach that will only find precise matches.
   *              It provides best init performance and provides good search performance.
   *              The 'fuzzyness' option will be ignored
   *
   * * 'fuzzy':   Search approach that implements a fuzzy (approximate) search.
   *              This search approach will find more results, even if there are no perfect matches.
   *              It has a moderate impact on init performance, and is slower when searching.
   *              It supports all options.
   *              Uses https://github.com/leeoniya/uFuzzy
   */
  searchStrategy: 'precise', // 'precise' or 'fuzzy'
  /**
   * Max search results. Reduce for better performance.
   * Does not apply for tag and folder search
   */
  searchMaxResults: 32,
  /**
   * Minimum string characters of the search term to consider a match
   */
  searchMinMatchCharLength: 1,
  /**
   * Fuzzy search threshold (0 - 1)
   * 0 is no fuzzyness, 1 is full fuzzyness
   *
   * This applies only to search approach 'fuzzy'.
   * For precise this is always 0 (no fuzzyness)
   */
  searchFuzzyness: 0.6,

  //////////////////////////////////////////
  // COLORS AND STYLE                     //
  //////////////////////////////////////////

  /**
   * Width of the left color marker in search results in pixels
   */
  colorStripeWidth: 6,

  /**
   * Color for bookmark results, expressed as CSS color
   */
  bookmarkColor: '#54afaf',
  /**
   * Color for tab results, expressed as CSS color
   */
  tabColor: '#b89aff',
  /**
   * Color for history results, expressed as CSS color
   */
  historyColor: '#9ece2f',
  /**
   * Color for search results, expressed as CSS color
   */
  searchColor: '#e1a535',
  /**
   * Color for custom search results, expressed as CSS color
   */
  customSearchColor: '#ce5c2f',

  //////////////////////////////////////////
  // SEARCH SOURCES                       //
  //////////////////////////////////////////

  /**
   * Whether to index and search for open browser tabs
   */
  enableTabs: true,
  /**
   * Whether to index and search for bookmarks
   */
  enableBookmarks: true,
  /**
   * Whether to index and search for browsing history
   * Please note that the history API tends to be slow,
   * so be careful about how many items you load.
   */
  enableHistory: true,
  /**
   * Enable or disable search engine links in results
   * Search engines can be a useful fallback mechanism to search externally for the term
   */
  enableSearchEngines: true,

  //////////////////////////////////////////
  // DISPLAY OPTIONS                      //
  //////////////////////////////////////////

  /**
   * Extract tags from title and display it as a badge with different search priority
   * Disabling this will also disable the tag overview and the tag search mode.
   */
  displayTags: true,
  /**
   * Display and search the folder names of bookmarks.
   * Disabling this will also disable the folder overview and the folder search mode.
   */
  displayFolderName: true,
  /**
   * Highlight search matches in results.
   * Reduces rendering performance a little.
   */
  displaySearchMatchHighlight: true,
  /**
   * Display last visit (time ago)
   */
  displayLastVisit: true,
  /**
   * Display how many times a result was visited (by history)
   */
  displayVisitCounter: false,
  /**
   * Display date when a bookmark was added
   */
  displayDateAdded: false,
  /**
   * Display result score.
   * The score indicates the relevance of the result and defines the order of results.
   */
  displayScore: true,

  //////////////////////////////////////////
  // TABS OPTIONS                         //
  //////////////////////////////////////////

  /**
   * If true, only the current browser window is considered for tab indexing and switching
   */
  tabsOnlyCurrentWindow: false,
  /**
   * When initializing search, show a certain number of tabs, sorted by last visited
   * Set this to 0 to disable.
   */
  tabsDisplayLastVisited: 0,
  /**
   * If true, the windowId of the tab result will be displayed
   * This can help to with multi-window situations
   * */
  tabsDisplayWindowId: false,

  //////////////////////////////////////////
  // HISTORY OPTIONS                      //
  //////////////////////////////////////////

  /**
   * How many days ago the browser history should be fetched
   */
  historyDaysAgo: 14,
  /**
   * How many history items should be fetched at most
   * Be careful, as too many items have negative impact on startup and search performance
   */
  historyMaxItems: 512,
  /**
   * All history items that start with the URLs given here will be skipped
   *
   * @example
   * historyIgnoreList: ["http://localhost"]
   */
  historyIgnoreList: [],

  //////////////////////////////////////////
  // SEARCH ENGINES OPTIONS               //
  //////////////////////////////////////////

  /**
   * Search Engine "Providers" which are used to pass on the search term via URL
   *
   * For each entry here, one result will be created - in the order they are defined.
   * The URLs need to include the search querystring (see examples).
   *
   * If an '$s' is part of the urlPrefix, it will be replaced with the search term.
   * Otherwise, the search term is appended to the urlPrefix.
   */
  searchEngineChoices: [
    {
      name: 'Google',
      urlPrefix: 'https://www.google.com/search?q=$s',
    },
    {
      name: 'Bing',
      urlPrefix: 'https://www.bing.com/search?q=$s',
    },
    {
      name: 'DuckDuckGo',
      urlPrefix: 'https://duckduckgo.com/?q=$s',
    },
    {
      name: 'dict.cc',
      urlPrefix: 'https://www.dict.cc/?s=$s',
    },
  ],

  /**
   * Allows to define custom search engines with their own custom alias
   * To trigger a search, type in the alias plus space: `<alias> ` to begin the search
   * The alias can be one string or an array of strings
   *
   * If an '$s' is part of the urlPrefix, it will be replaced with the search term.
   * Otherwise, the search term is appended to the urlPrefix.
   *
   * Optionally, a "blank" can be given, which is the URL chosen when there is no search string.
   * With this, they act like a "high-priority" custom bookmark.
   */
  customSearchEngines: [
    {
      alias: ['g', 'google'],
      name: 'Google',
      urlPrefix: 'https://www.google.com/search?q=$s',
      blank: 'https://www.google.com',
    },
    {
      alias: ['d', 'dict'],
      name: 'dict.cc',
      urlPrefix: 'https://www.dict.cc/?s=$s',
      blank: 'https://www.dict.cc',
    },
  ],

  //////////////////////////////////////////
  // SCORE CALCULATION OPTIONS            //
  //////////////////////////////////////////

  /**
   * Filter out all search results below this minimum score
   */
  scoreMinScore: 30,
  /**
   * Minimum ratio of search term matches to consider a match.
   * Set to 1 to only return results that match all search terms
   * Set to 0 to return all results that match at least one search term, although with reduced score
   *
   * This setting only applies to precise search
   */
  scoreMinSearchTermMatchRatio: 0.6,

  // RESULT TYPE BASE SCORES
  // Depending on the type of result, they start with a base score
  // Please make sure that this is not below the minScore :)
  /**
   * Base score for bookmark results
   */
  scoreBookmarkBaseScore: 100,
  /**
   * Base score for tab results
   */
  scoreTabBaseScore: 70,
  /**
   * Base score for history results
   */
  scoreHistoryBaseScore: 50,
  /**
   * Base score for search engine choices
   */
  scoreSearchEngineBaseScore: 30,
  /**
   * Base score for custom search engine choices
   * This is set very high to ensure that it's the topmost entry
   */
  scoreCustomSearchEngineBaseScore: 500,

  // FIELD WEIGHTS
  // Depending on in which field the search match was found,
  // the match gets a multiplier applied on how important the match is.
  /** Weight for a title match*/
  scoreTitleWeight: 1,
  /** Weight for a tag match*/
  scoreTagWeight: 0.7,
  /** Weight for an url match*/
  scoreUrlWeight: 0.6,
  /** Weight for a folder match*/
  scoreFolderWeight: 0.5,

  // BONUS SCORES
  // If certain conditions apply, extra score points can be added
  /**
   * If enabled, bookmarks can add custom bonus scores by putting it in the title
   * HowTo: Add a ` +<score>` (space, plus sign, whole number) in the title before tags.
   * Do not add bonus scores more than once in a single bookmark title.
   * E.g: `Bookmark Title +20` or `Another Bookmark +10 #tag1 #tag2`
   * */
  scoreCustomBonusScore: true,
  /**
   * For each exact "includes" match we add some bonus points
   */
  scoreExactIncludesBonus: 5,
  /**
   * The minimum characters a search term needs to have to consider a exact includes match
   */
  scoreExactIncludesBonusMinChars: 3,
  /**
   * Additional score points if title or url starts exactly with the search text.
   * This comes on top of an include bonus.
   */
  scoreExactStartsWithBonus: 10,
  /**
   * Additional score points if title matches exactly with the search text.
   * This comes on top of an include and starts with bonus.
   */
  scoreExactEqualsBonus: 15,
  /**
   * Additional points for an exact match of a search term tag
   */
  scoreExactTagMatchBonus: 10,
  /**
   * Additional points for an exact match of a search term folder name
   */
  scoreExactFolderMatchBonus: 5,
  /**
   * Adds score points for every site visit according to browsing history
   * Please note that only history items within `history.daysAgo` can be considered,
   * however the visited counter itself considers your complete history.
   */
  scoreVisitedBonusScore: 0.25,
  /**
   * Maximum score points for visited bonus
   */
  scoreVisitedBonusScoreMaximum: 10,
  /**
   * Adds score points when a bookmark or history has been accessed recently.
   * Calculated by taking the recentBonusScoreMaximum and subtract recentBonusScorePerHour
   * for each hour the access happened in the past.
   * There is no negative score.
   *
   * Example: If maximum is 24 and perHour is 0.5:
   * * For a page just opened there will be ~20 bonus score
   * * For a page opened 24 hours ago there will be 10 bonus score
   * * For a page opened 48 hours ago there will be 0 bonus score
   */
  scoreRecentBonusScorePerHour: 0.5,
  scoreRecentBonusScoreMaximum: 20,
  /**
   * Adds score points when a bookmark has been added more recently.
   * Calculated by taking the dateAddedBonusScoreMaximum and subtract dateAddedBonusScorePerDay
   * for each day the bookmark has been added in the past.
   * There is no negative score.
   */
  scoreDateAddedBonusScorePerDay: 0.1,
  scoreDateAddedBonusScoreMaximum: 5,

  //////////////////////////////////////////
  // POWER USER OPTIONS                   //
  //////////////////////////////////////////

  // Those are only meant for power users who know what they're doing
  // And those options may also not be long-time stable

  /**
   * If the extension detects that the title is just the URL of the link,
   * it gets shortened so the result entries are not swamped by long URL patterns
   */
  titleLengthRestrictionForUrls: 80,

  /**
   * Customized options for the fuzzy search library uFuzzy ('@leeoniya/ufuzzy')
   * @see https://github.com/leeoniya/uFuzzy/blob/main/src/uFuzzy.js#L9
   */
  uFuzzyOptions: {},
}

/**
 * If there are no options yet, use this as an empty options template
 */
export const emptyOptions = {
  searchStrategy: defaultOptions.searchStrategy,
}

/**
 * Writes user settings to the google chrome sync storage
 *
 * @see https://developer.chrome.com/docs/extensions/reference/storage/
 */
export async function setUserOptions(userOptions) {
  return new Promise((resolve, reject) => {
    userOptions = userOptions || {}

    try {
      validateUserOptions(userOptions)
    } catch (err) {
      printError(err, 'Could not save user options.')
      return reject(err)
    }

    if (ext.browserApi.storage) {
      ext.browserApi.storage.sync.set({ userOptions: userOptions }, () => {
        if (ext.browserApi.runtime.lastError) {
          return reject(ext.browserApi.runtime.lastError)
        }
        return resolve()
      })
    } else {
      console.warn('No storage API found. Falling back to local Web Storage')
      window.localStorage.setItem('userOptions', JSON.stringify(userOptions))
      return resolve()
    }
  })
}

/**
 * Get user options.
 * If none are stored yet, this will return the default empty options
 */
export async function getUserOptions() {
  return new Promise((resolve, reject) => {
    try {
      if (ext.browserApi.storage && ext.browserApi.storage.sync && ext.browserApi.storage.sync.get) {
        ext.browserApi.storage.sync.get(['userOptions'], (result) => {
          if (ext.browserApi.runtime.lastError) {
            return reject(ext.browserApi.runtime.lastError)
          }
          const userOptions = migrateOptions(result.userOptions || emptyOptions)
          return resolve(userOptions)
        })
      } else {
        console.warn('No storage API found. Falling back to local Web Storage')
        const userOptionsString = window.localStorage.getItem('userOptions')
        const userOptions = migrateOptions(userOptionsString ? JSON.parse(userOptionsString) : emptyOptions)
        return resolve(userOptions)
      }
    } catch (err) {
      return reject(err)
    }
  })
}

/**
 * Gets the actual effective options based on the default options
 * and the overrides of the user options
 */
export async function getEffectiveOptions() {
  try {
    const userOptions = await getUserOptions()
    validateUserOptions(userOptions)
    return {
      ...defaultOptions,
      ...userOptions,
    }
  } catch (err) {
    printError(err, 'Could not get valid user options, falling back to defaults.')
    return defaultOptions
  }
}

export function validateUserOptions(userOptions) {
  if (userOptions) {
    if (typeof userOptions !== 'object') {
      throw new Error('User options must be a valid YAML / JSON object')
    }
    try {
      JSON.stringify(userOptions)
    } catch (err) {
      throw new Error('User options cannot be parsed into JSON: ' + err.message)
    }
  }
}

function migrateOptions(userOptions) {
  if (userOptions && userOptions.searchStrategy === 'hybrid') {
    userOptions.searchStrategy = 'fuzzy'
  }
  return userOptions
}
